sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/utils/SelectHelpersTest.java

```diff
@@ -0,0 +1,345 @@
+
+    Row row = SelectHelpers.selectRow(MAP_ROW, fieldAccessDescriptor, MAP_SCHEMA, outputSchema);
+
+    Row expectedValueRow = Row.withSchema(expectedValueSchema).addValue("first").build();
+    Row expectedRow =
+        Row.withSchema(expectedSchema).addValue(ImmutableMap.of(1, expectedValueRow)).build();
+    assertEquals(expectedRow, row);
+  }
+
+  @Test
+  public void testSelectMapOfRowSelectAll() {
+    FieldAccessDescriptor fieldAccessDescriptor =
+        FieldAccessDescriptor.withFieldNames("map{}.*").resolve(MAP_SCHEMA);
+    Schema outputSchema = SelectHelpers.getOutputSchema(MAP_SCHEMA, fieldAccessDescriptor);
+    Schema expectedSchema =
+        Schema.builder().addMapField("map", FieldType.INT32, FieldType.row(FLAT_SCHEMA)).build();
+    assertEquals(expectedSchema, outputSchema);
+
+    Row row = SelectHelpers.selectRow(MAP_ROW, fieldAccessDescriptor, MAP_SCHEMA, outputSchema);
+    Row expectedRow = Row.withSchema(expectedSchema).addValue(ImmutableMap.of(1, FLAT_ROW)).build();
+    assertEquals(expectedRow, row);
+  }
+
+  @Test
+  public void testSelectMapOfArray() {
+    FieldAccessDescriptor fieldAccessDescriptor =
+        FieldAccessDescriptor.withFieldNames("map.field1").resolve(MAP_ARRAY_SCHEMA);
+    Schema outputSchema = SelectHelpers.getOutputSchema(MAP_ARRAY_SCHEMA, fieldAccessDescriptor);
+
+    Schema expectedValueSchema = Schema.builder().addStringField("field1").build();
+    Schema expectedSchema =
+        Schema.builder()
+            .addMapField(
+                "map", FieldType.INT32, FieldType.array(FieldType.row(expectedValueSchema)))
+            .build();
+    assertEquals(expectedSchema, outputSchema);
+
+    Row row =
+        SelectHelpers.selectRow(
+            MAP_ARRAY_ROW, fieldAccessDescriptor, MAP_ARRAY_SCHEMA, outputSchema);
+    Row expectedElement = Row.withSchema(expectedValueSchema).addValue("first").build();
+
+    Row expectedRow =
+        Row.withSchema(expectedSchema)
+            .addValue(ImmutableMap.of(1, ImmutableList.of(expectedElement)))
+            .build();
+    assertEquals(expectedRow, row);
+  }
+
+  @Test
+  public void testSelectFieldOfRecord() {
+    Schema f1 = Schema.builder().addInt64Field("f0").build();
+    Schema f2 = Schema.builder().addRowField("f1", f1).build();
+    Schema f3 = Schema.builder().addRowField("f2", f2).build();
+
+    Row r1 = Row.withSchema(f1).addValue(42L).build(); // {"f0": 42}
+    Row r2 = Row.withSchema(f2).addValue(r1).build(); // {"f1": {"f0": 42}}
+    Row r3 = Row.withSchema(f3).addValue(r2).build(); // {"f2": {"f1": {"f0": 42}}}
+
+    FieldAccessDescriptor fieldAccessDescriptor =
+        FieldAccessDescriptor.withFieldNames("f2.f1").resolve(f3);
+
+    Schema outputSchema = SelectHelpers.getOutputSchema(f3, fieldAccessDescriptor);
+
+    Row out = SelectHelpers.selectRow(r3, fieldAccessDescriptor, r3.getSchema(), outputSchema);
+
+    assertEquals(outputSchema, f2);
+    assertEquals(out, r2);
```

Comment:
Parameters should be in a different order: expected, actual

Reply:
done